src/core/graph.test.js

```diff
@@ -1358,6 +1358,58 @@ describe("core/graph", () => {
     });
   });
 
+  describe("contract", () => {
+    const a = node("a");
+    const b = node("b");
+    const c = node("c");
+    it("has no effect with no contractions", () => {
+      const g = simpleGraph();
+      const g_ = simpleGraph().contract([]);
+      expect(g.equals(g_)).toBe(true);
+    });
+    it("adds the new node to the graph", () => {
+      const g = simpleGraph().contract([{old: [], replacement: c}]);
+      const g_ = simpleGraph().addNode(c);
+      expect(g.equals(g_)).toBe(true);
+    });
+    it("filters old nodes from the graph", () => {
+      const g = new Graph()
+        .addNode(a)
+        .addNode(b)
+        .contract([{old: [a.address, b.address], replacement: c}]);
+      const expected = new Graph().addNode(c);
+      expect(g.equals(expected)).toBe(true);
+    });
+    it("re-writes edges, including dangling or loop edges", () => {
+      const g = new Graph()
+        .addNode(a)
+        .addEdge(edge("loop", a, a))
+        .addEdge(edge("dangle1", a, b))
+        .addEdge(edge("dangle2", b, a))
+        .contract([{old: [a.address], replacement: c}]);
```

Comment:
Maybe also test a normal (non-dangling, non-loop) edge, in- and out-?

Reply:
Done.